File: src/app/services/navigation.service.ts
import { Injectable } from '@angular/core';

export class RoutePoint {
  constructor(
    public latitude: number,
    public longitude: number,
    public altitude: number
  ) { }

  get LngLat() {
    return [this.longitude, this.latitude];
  }
}

export class NewRouteEvent extends Event {
  constructor(
    public route: RoutePoint[],
    public name: string
  ) {
    super("newRoute");
  }
}

@Injectable({
  providedIn: 'root'
})
export class NavigationService {

  private route: RoutePoint[] = [];
  private routeName: string = '';
  protected newRouteTarget: EventTarget = new EventTarget();;


  loadRouteFileGPX(content: string) {
    this.route = [];
    const parser = new DOMParser();
    const xmlDoc = parser.parseFromString(content, "text/xml");
    this.routeName = xmlDoc.getElementsByTagName("name")[0].textContent || 'GPX Route';
    const trkpts = xmlDoc.getElementsByTagName("trkpt");
    for (let i = 0; i < trkpts.length; i++) {
      const lat = parseFloat(trkpts[i].getAttribute("lat")!);
      const lon = parseFloat(trkpts[i].getAttribute("lon")!);
      const ele = parseFloat(trkpts[i].getElementsByTagName("ele")[0].textContent!);
      this.route.push(new RoutePoint(lat, lon, ele));
    }
    this.newRouteTarget.dispatchEvent(new NewRouteEvent(this.route, this.routeName))
  }

  addNewRouteListener(callback: (event: NewRouteEvent) => void) {
    this.newRouteTarget.addEventListener('newRoute', callback as any);
  }
  removeNewRouteListener(callback: (event: NewRouteEvent) => void) {
    this.newRouteTarget.removeEventListener('newRoute', callback as any);
  }

  hasRoute(): boolean {
    return this.route && this.route.length > 0;
  }

  getRouteName(): string {
    return this.routeName || '';
  }
}

File: android/app/src/main/java/com/openbikecomputer/service/NavigationService.java
package com.openbikecomputer.service;

import android.util.Xml;

import com.openbikecomputer.model.RoutePoint;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;
import java.io.StringReader;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * Loads and stores a GPX route. Ported from {@code navigation.service.ts}.
 */
public class NavigationService {

    /** Listener notified when a new route is loaded. */
    public interface OnNewRoute {
        void onNewRoute(List<RoutePoint> route, String name);
    }

    private List<RoutePoint> route = new ArrayList<>();
    private String routeName = "";
    private final CopyOnWriteArrayList<OnNewRoute> listeners = new CopyOnWriteArrayList<>();

    /** Parse GPX content and store the contained track. */
    public void loadRouteFileGPX(String content) {
        List<RoutePoint> parsed = new ArrayList<>();
        String name = "GPX Route";
        try {
            XmlPullParser parser = Xml.newPullParser();
            parser.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false);
            parser.setInput(new StringReader(content));

            double lat = 0, lon = 0, ele = 0;
            boolean inTrkpt = false;
            boolean nameCaptured = false;
            String currentTag = null;

            int event = parser.getEventType();
            while (event != XmlPullParser.END_DOCUMENT) {
                if (event == XmlPullParser.START_TAG) {
                    String tag = parser.getName();
                    currentTag = tag;
                    if ("trkpt".equalsIgnoreCase(tag)) {
                        inTrkpt = true;
                        ele = 0;
                        String latAttr = parser.getAttributeValue(null, "lat");
                        String lonAttr = parser.getAttributeValue(null, "lon");
                        lat = latAttr != null ? Double.parseDouble(latAttr) : 0;
                        lon = lonAttr != null ? Double.parseDouble(lonAttr) : 0;
                    }
                } else if (event == XmlPullParser.TEXT) {
                    String text = parser.getText();
                    if (text != null) {
                        text = text.trim();
                        if (!text.isEmpty() && currentTag != null) {
                            if (!nameCaptured && "name".equalsIgnoreCase(currentTag)) {
                                name = text;
                                nameCaptured = true;
                            } else if (inTrkpt && "ele".equalsIgnoreCase(currentTag)) {
                                ele = Double.parseDouble(text);
                            }
                        }
                    }
                } else if (event == XmlPullParser.END_TAG) {
                    String tag = parser.getName();
                    if ("trkpt".equalsIgnoreCase(tag)) {
                        parsed.add(new RoutePoint(lat, lon, ele));
                        inTrkpt = false;
                    }
                    currentTag = null;
                }
                event = parser.next();
            }
        } catch (XmlPullParserException | IOException | NumberFormatException e) {
            throw new RuntimeException("Failed to parse GPX route", e);
        }

        this.route = parsed;
        this.routeName = name;
        for (OnNewRoute listener : listeners) {
            listener.onNewRoute(this.route, this.routeName);
        }
    }

    public void addNewRouteListener(OnNewRoute listener) {
        listeners.add(listener);
    }

    public void removeNewRouteListener(OnNewRoute listener) {
        listeners.remove(listener);
    }

    public boolean hasRoute() {
        return route != null && !route.isEmpty();
    }

    public String getRouteName() {
        return routeName != null ? routeName : "";
    }

    public List<RoutePoint> getRoute() {
        return route;
    }
}
